src/System.IO.Pipelines/BufferSegment.cs

```diff
@@ -128,16 +133,37 @@ public static BufferSegment Clone(ReadCursor beginBuffer, ReadCursor endBuffer,
 
             while (beginOrig != endOrig)
             {
-                endClone.Next = new BufferSegment(beginOrig._owned, beginOrig.Start, beginOrig.End);
+                endClone.SetNext(new BufferSegment(beginOrig._owned, beginOrig.Start, beginOrig.End));
 
                 endClone = endClone.Next;
                 beginOrig = beginOrig.Next;
             }
 
             lastSegment = new BufferSegment(endOrig._owned, endOrig.Start, endBuffer.Index);
-            endClone.Next = lastSegment;
+            endClone.SetNext(lastSegment);
 
             return beginClone;
         }
+
+        public void SetNext(BufferSegment segment)
+        {
+            if (segment == null)
+            {
+                throw new ArgumentNullException(nameof(segment));
+            }
+            if (Next != null)
+            {
+                throw new InvalidOperationException($"{nameof(Next)} is already set.");
+            }
```

Comment:
This is only called internally. When we were just calling `segment.Next = nextSegment` there weren't any of these silly safety checks. I'm fine with making these asserts.

@benaadams Is this a good candidate for the aggressive inlining attribute?